sql/metadata.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Marc Berhault (marc@cockroachlabs.com)
+
+package sql
+
+import (
+	"sort"
+
+	"github.com/cockroachdb/cockroach/keys"
+	"github.com/cockroachdb/cockroach/roachpb"
+	"github.com/cockroachdb/cockroach/util/log"
+)
+
+// MetadataSchema is used to construct the initial sql schema for a new
+// CockroachDB cluster being bootstrapped. Tables and databases must be
+// installed on the underlying persistent storage before a cockroach store can
+// start running correctly, thus requiring this special initialization.
+type MetadataSchema struct {
+	systemObjects []systemObject
+	databases     []*MetadataDatabase
+}
+
+type systemObject struct {
+	parentID ID
+	desc     descriptorProto
+}
+
+// MetadataDatabase represents a database to be created on a bootstrapped
+// cockroach cluster. This structure should only be created by calling the
+// "AddDatabase()" method of a MetadataSchema object.
+type MetadataDatabase struct {
+	name       string
+	privileges *PrivilegeDescriptor
+	tables     []*metadataTable
+}
+
+type metadataTable struct {
+	definition string
+	privileges *PrivilegeDescriptor
+}
+
+// NewMetadataSchema constructs a new MetadataSchema object which contains no
+// tables.
+func NewMetadataSchema() *MetadataSchema {
```

Comment:
why does this need to return a pointer? actually, why bother with a constructor at all?